backend/audit/views/views.py

```diff
@@ -443,6 +448,10 @@
                 "submission_status": sac.submission_status,
             }
 
+            # Return to checklist, the Audit is not in the correct state.
+            if SingleAuditChecklist.STATUS.READY_FOR_CERTIFICATION not in sac.transition_name:
+                return redirect(f"/audit/submission-progress/{report_id}")
```

Comment:
I think the CodeQL warning here is safe to ignore. The `report_id` being used in a URL construction... well, it probably isn't an issue. If it comes in via the URL, then perhaps some kind of check should be made on the value before redirecting to it. (Do we have a `is_valid_report_id` function laying around anywhere?)

The question I have:

Is it enough to check that `READY_FOR_CERTIFICATION` is in the list of `transition_name`s? Because users can go back-and-forth between states, will this code succeed if the user was previously in `READY_FOR_CERTIFICATION`, backed out, and is now in a... not-ready state? That is, if the states are ABC, RFC, and DEF are possible states...

[ ABC ] will redirect back

[ ABC -> RFC ] will not.

[ ABC -> RFC -> ABC] will not.

Asking because I haven't looked around this part of the code for quite a while.

Reply:
@jadudm I cannot seem to find anything akin to a `is_valid_report_id` but I will keep searching around. However, using an invalid `report_id` in the URL returns a 403 webpage because each of these view methods begin with `SingleAuditChecklist.objects.get(report_id=...)`, which will throw a 403 webpage on no records or failure. This guarantees that the redirect will only occur if the report exists.

And that is a great second point. Using the Auditor Certification step as an example, let's say I do the following:

1. I have finished uploading my workbooks and run pre-submission validation. This allows me to lock my audit to begin Auditor Certification.
2. I can now access `/audit/auditor-certification-confirm/<report_id>` since `READY_FOR_CERTIFICATION` is in my transition_name list.
3. I COULD unlock the audit right after this, which would make my `transition_name` look like:
   ```
   [ READY_FOR_CERTIFICATION -> IN_PROGRESS ]
   ```
4. If I did this, I could still access `/audit/auditor-certification-confirm/<report_id>`. This is the problem.

I can create a class method in SingleAuditChecklist that returns the latest state in `transition_name`. From my own testing it seems that the most recent state always appears in the last index. Is it safe for me to assume this is true, or should I check against the `transition_date` list to guarantee the most recent state for the audit?

Reply:
I think if it is the latest state/last index, we'll be OK. 

At the moment, we only append to those state lists, and if we ever break that assumption, we're in trouble for lots of reasons. So, looking at the last index to see if we're in the right state for this feels right to me.